### PR TITLE
Feat: add support for a custom route prefix

### PR DESCRIPTION
**Pull request by sharafian**
adds an extra argument to `addRoute` allowing a custom destination prefix to be specified. Depended on by https://github.com/interledgerjs/ilp-connector/pull/238

### CODE REVIEW

File: src/lib/routing-table.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ class RoutingTable {
     this.destinations = new PrefixMap()
   }
 
-  addRoute (destination, nextHop, route) {
+  addRoute (destination, nextHop, route, prefix) {
```
Comment (sentientwaffle) [inline src/lib/routing-table.js]: Why is this here?

File: test/routing-tables.test.js [hunk 2]
```diff
@@ -122,6 +123,26 @@ describe('RoutingTables', function () {
         undefined)
     })
 
+    it('creates a route with a custom prefix, if supplied', function () {
+      const route = Route.fromData({
```
Comment (sentientwaffle) [inline test/routing-tables.test.js]: `addRoute` will call `Route.fromData`, you don't need to do it here.

File: src/lib/routing-tables.js [hunk 3]
```diff
@@ -56,7 +56,7 @@ class RoutingTables {
    * @param {Route|RouteData} _route from ledger B→C
    * @returns {Boolean} whether or not a new route was added
    */
-  addRoute (_route) {
+  addRoute (_route, prefix) {
```
Comment (sentientwaffle) [inline src/lib/routing-tables.js]: Just calling this `prefix` isn't very descriptive of _which_ prefix, especially since any ledger address is actually a prefix as well.

Comment (sentientwaffle) [inline src/lib/routing-tables.js]: It might also be simpler to add it as a new property of `Route` rather than passing it around separately.

File: test/routing-tables.test.js (hunk 2 above)
Comment (sharafian) [inline test/routing-tables.test.js]: removed the `Route.fromData` call

File: src/lib/routing-tables.js (hunk 3 above)
Comment (sharafian) [inline src/lib/routing-tables.js]: renamed to `destinationPrefix`. I think it might be better to keep it separate from the route object, because in almost all cases it would be redundant and cause problems if it were not assigned

Comment (sentientwaffle) [inline src/lib/routing-tables.js]: Now that I think about it, would `targetPrefix` be a better name? I think it would be helpful for differentiating between that and `destinationLedger`.

Comment (sharafian) [inline src/lib/routing-tables.js]: I like destinationPrefix because it's the key added to the `PrefixMap` named `destinations`. The route constructor already defaults destinationPrefix to destinationLedger, I believe: https://github.com/interledgerjs/ilp-routing/pull/13/files#diff-461b64a626e26d6495dc1eed9f8787d5R92

File: src/lib/route.js [hunk 4]
```diff
@@ -21,6 +21,11 @@ class Route {
     this.nextLedger = hops[1]
     this.destinationLedger = hops[hops.length - 1]
 
+    // if targetPrefix is specified, then destinations matching 'targetPrefix'
+    // will follow this route, rather than destinations matching
+    // 'destinationLedger'
+    this.targetPrefix = info.targetPrefix || this.destinationLedger
```
Comment (sentientwaffle) [inline src/lib/route.js]: Add `info.targetPrefix` to the constructor's jsdoc.

File: src/lib/routing-tables.js [hunk 5]
```diff
@@ -66,13 +66,13 @@ class RoutingTables {
     this.eachSource((tableFromA, ledgerA) => {
       added = this._addRouteFromSource(tableFromA, ledgerA, route) || added
     })
-    if (added) debug('add route', route.sourceAccount, route.destinationLedger)
+    if (added) debug('add routem matching', route.targetPrefix, ':', route.sourceAccount, route.destinationLedger)
```
Comment (sentientwaffle) [inline src/lib/routing-tables.js]: `route matching`, not `routem matching`

File: test/routing-tables.test.js [hunk 6]
```diff
@@ -120,6 +120,24 @@ describe('RoutingTables', function () {
...
+        min_message_window: 1,
+        points: [ [0, 0], [50, 60] ]
+      }
+
+      // will be stored in destinations under 'prefix.'
+      assert.equal(this.tables.addRoute(route), true)
+
+      assertSubset(
+        this.tables.sources.get(ledgerA).destinations.get('prefix.').get(ledgerC + 'mary'),
+        { destinationLedger: ledgerC })
```
Comment (sentientwaffle) [inline test/routing-tables.test.js]: Verify `targetPrefix` as well.

File: src/lib/route.js (hunk 4 above)
Comment (sharafian) [inline src/lib/route.js]: fixed

File: test/routing-tables.test.js (hunk 6 above)
Comment (sharafian) [inline test/routing-tables.test.js]: In this case, the `targetPrefix` is not preserved in the final route due to joining the routes (although it still is used for the entry in the destinations table). Should `Route.join` preserve custom route prefixes? If you joined two routes with custom `targetPrefix`es, how do you decide which to preserve in the final route?

Comment (sharafian) [inline test/routing-tables.test.js]: alright, I set it so `join` uses the `tailRoute`'s target prefix, and I test for it in the custom prefix test
